### PR TITLE
Allow DefaultHttpContext to be pooled [Design]

### PR DESCRIPTION
**Pull request by benaadams**
Half of  #402 needs change to Hosting also

### CODE REVIEW

File: src/Microsoft.AspNet.Http/DefaultHttpContext.cs [hunk 1]
```diff
@@ -168,8 +185,26 @@ public override void Abort()
 
         public override void Dispose()
         {
-            // REVIEW: is this necessary? is the environment "owned" by the context?
```
Comment (benaadams) [inline src/Microsoft.AspNet.Http/DefaultHttpContext.cs]: Question has moved to -> https://github.com/aspnet/HttpAbstractions/pull/404/files#diff-ad404d59bc5860d71b7403b8e0a665b2R102

File: src/Microsoft.AspNet.Http.Features/FeatureReference.cs [hunk 2]
```diff
@@ -34,5 +34,11 @@ public T Update(IFeatureCollection features, T feature)
             _revision = features.Revision;
             return feature;
         }
+
+        public void Reset()
```
Comment (davidfowl) [inline src/Microsoft.AspNet.Http.Features/FeatureReference.cs]: Do you really need to hold onto FeatureReference structs?

File: src/Microsoft.AspNet.Http/Authentication/DefaultAuthenticationManager.cs [hunk 3]
```diff
@@ -13,14 +13,19 @@
 
 namespace Microsoft.AspNet.Http.Authentication.Internal
 {
-    public class DefaultAuthenticationManager : AuthenticationManager
+    public class DefaultAuthenticationManager : AuthenticationManager, IDisposable
```
Comment (davidfowl) [inline src/Microsoft.AspNet.Http/Authentication/DefaultAuthenticationManager.cs]: I'm not a fan of using IDisposable to release memory for the pooled object, feels like mismatch...

File: src/Microsoft.AspNet.Http/Authentication/DefaultAuthenticationManager.cs [hunk 4]
```diff
@@ -109,5 +114,13 @@ public override IEnumerable<AuthenticationDescription> GetAuthenticationSchemes(
                 throw new InvalidOperationException($"The following authentication scheme was not accepted: {authenticationScheme}");
             }
         }
+
+        public void Dispose()
```
Comment (davidfowl) [inline src/Microsoft.AspNet.Http/Authentication/DefaultAuthenticationManager.cs]: Reset

File: src/Microsoft.AspNet.Http/DefaultConnectionInfo.cs [hunk 5]
```diff
@@ -72,5 +78,13 @@ public override Task<X509Certificate2> GetClientCertificateAsync(CancellationTok
         {
             return TlsConnectionFeature.GetClientCertificateAsync(cancellationToken);
         }
+
+        public void Dispose()
```
Comment (davidfowl) [inline src/Microsoft.AspNet.Http/DefaultConnectionInfo.cs]: Reset

File: src/Microsoft.AspNet.Http/DefaultHttpContext.cs [hunk 6]
```diff
@@ -51,6 +58,16 @@ public DefaultHttpContext(IFeatureCollection features)
             _session = FeatureReference<ISessionFeature>.Default;
         }
 
+        public void Reinitalize(IFeatureCollection features)
```
Comment (davidfowl) [inline src/Microsoft.AspNet.Http/DefaultHttpContext.cs]: Initialize

File: src/Microsoft.AspNet.Http/DefaultHttpContext.cs [hunk 7]
```diff
@@ -168,8 +185,26 @@ public override void Abort()
 
         public override void Dispose()
```
Comment (davidfowl) [inline src/Microsoft.AspNet.Http/DefaultHttpContext.cs]: Reset

**Comment by benaadams** [comment]
See https://github.com/aspnet/HttpAbstractions/issues/402#issuecomment-140336990 for details on server run improvements (61.5k rps -> 1.49M rps)

File: src/Microsoft.AspNet.Http/Authentication/DefaultAuthenticationManager.cs (hunk 3 above)
Comment (benaadams) [inline src/Microsoft.AspNet.Http/Authentication/DefaultAuthenticationManager.cs]: Perhaps IReusable? With .Clear() method?

File: src/Microsoft.AspNet.Http.Features/FeatureReference.cs (hunk 2 above)
Comment (benaadams) [inline src/Microsoft.AspNet.Http.Features/FeatureReference.cs]: Need to check; they are heap allocated - so would be another chunk of memory but they are only a pointer and an int...

File: src/Microsoft.AspNet.Http/DefaultHttpResponse.cs [hunk 8]
```diff
@@ -116,5 +121,14 @@ public override void Redirect(string location, bool permanent)
 
             Headers[HeaderNames.Location] = location;
         }
+
+        public void Dispose()
+        {
```
Comment (benaadams) [inline src/Microsoft.AspNet.Http/DefaultHttpResponse.cs]: Reset

File: src/Microsoft.AspNet.Http/DefaultHttpContext.cs (hunk 7 above)
Comment (benaadams) [inline src/Microsoft.AspNet.Http/DefaultHttpContext.cs]: Only one that doesn't fit as is already Disposable?

**Comment by davidfowl** [comment]
Closing this as discussed with @benaadams

**Pull request closed without merging**